frost-core/src/frost/identifier.rs

```diff
@@ -14,73 +14,132 @@ use crate::{Ciphersuite, Error, Field, Group, Scalar};
 /// over, corresponding to some x-coordinate for a polynomial f(x) = y.  MUST NOT be zero in the
 /// field, as f(0) = the shared secret.
 #[derive(Copy, Clone)]
-pub struct Identifier<C>(u16, PhantomData<C>);
+pub struct Identifier<C: Ciphersuite>(pub(crate) Scalar<C>);
 
-impl<C> Identifier<C>
+impl<C> Debug for Identifier<C>
 where
     C: Ciphersuite,
 {
-    // Convert the identifier to a Scalar.
-    pub(crate) fn to_scalar(self) -> Scalar<C> {
-        // Classic left-to-right double-and-add algorithm that skips the first bit 1 (since
-        // identifiers are never zero, there is always a bit 1), thus `sum` starts with 1 too.
-        let one = <<C::Group as Group>::Field as Field>::one();
-        let mut sum = <<C::Group as Group>::Field as Field>::one();
-        let bits = (self.0.to_be_bytes().len() as u32) * 8;
-        for i in (0..(bits - self.0.leading_zeros() - 1)).rev() {
-            sum = sum + sum;
-            if self.0 & (1 << i) != 0 {
-                sum = sum + one;
-            }
-        }
-        sum
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_tuple("Identifier")
+            .field(&usize::from(*self))
+            .finish()
     }
 }
 
-impl<C> PartialEq for Identifier<C> {
-    fn eq(&self, other: &Self) -> bool {
-        self.0 == other.0
+impl<C> Eq for Identifier<C> where C: Ciphersuite {}
+
+impl<C> From<Identifier<C>> for u16
+where
+    C: Ciphersuite,
+{
+    fn from(id: Identifier<C>) -> u16 {
+        let mut bytes = [0u8; 2];
+
+        let serialized = <<C::Group as Group>::Field as Field>::serialize(&id.0);
+
+        bytes.copy_from_slice(&serialized.as_ref()[..2]);
+
+        u16::from_le_bytes(bytes)
```

Comment:
Unfortunately this does not always work because there is no guarantee that the scalar is serialized in little endian (which was the motivation for this [PR](https://github.com/cfrg/draft-irtf-cfrg-frost/pull/229))

It's passing CI here because our tests don't use indices larger than 1 byte, but it will fail in that case. I pushed a test vector that demonstrates this in 9a3ffda3bea4aa343bcc2b1a7ccd8093efa661f8

Reply:
funny nit, encode_u16 is big endian ![Screenshot_20220823-100843.png](https://user-images.githubusercontent.com/552961/186180362-dc998e6d-154c-4903-95b7-072e71e2341b.png)

Reply:
That depends on how you interpret the spec... if you consider that the identifier is an "integer", then you need `integer -> Scalar` and `integer -> encoded u16`. The first one is doable if a bit awkward and the second is easy.

If you consider that an identifier is a Scalar, then you will still need `integer -> Scalar` because you will need to build the scalar somehow, and  `Scalar -> encoded u16`. And I don't see a ciphersuite-generic way to do the `Scalar -> encoded u16` conversion. What you must guarantee is that the index 1 is mapped to the Scalar 1 and so on, which is not happening here. I feel that this second approach is trying to force a leaky abstraction somehow.

Reply:
identifiers are created and operated over as scalar field elements during key gen operations over the secret polynomial, so 'at heart' they are scalars that we arbitrarily cap in value at MAX_SIGNERS. I do think the spec is not doing the Scalar <-> uint16 enforcement any favors here, and maybe we need to define a per-ciphersuite routine for this or a generic routine _for internal operations only_ (not deserialization/serialization b/c p256, etc) to make this explicit

Reply:
I'm not sure. Like I mentioned the spec defines identifier as integers, which are now explicitly converted to scalar with `Scalar(i)` in the spec. I wouldn't mind either way, but I don't see the problem with the current approach of using u16:

- It doesn't require the Scalar -> encoded u16 convertion, which does not seem possible to do in a ciphersuite-generic way without a new trait method
- Seems functionally equal in every other aspect (maybe there is a small performance penalty of converting multiple times, but it seems negligible, and could be improved by storing both u16 and Scalar, if needed)

Reply:
(We decided to go ahead and define identifiers as scalar, including in the spec)